### PR TITLE
LVRUBYM-217, LVRUBYM-218: Apply authorization logic to the API

### PR DESCRIPTION
**Pull request by vikabelous**
## JIRA

* [Main JIRA ticket](http://ssu-jira.softserveinc.com/browse/LVRUBYM-204)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-217)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-218)

## Code reviewers

- [x] @skrypnyk88 
- [x] @JulkaIII 
- [x] @Mykhailo123 

### Second Level Review

- [ ] @omaik

## Summary of issue

Need to implement authentication and authorization logic with JWT

## Summary of change

- added JWT and Rack-Cors gems
- added AuthenticationController
- added before filter for ApplicationController
- added JWT authorization strategy for Warden
- updated StudentsController specs
- added AuthenticationController specs

## CHECK LIST
- [x]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: config/initializers/devise.rb [hunk 1]
```diff
@@ -29,4 +29,44 @@
...
   config.sign_out_via = :delete
+
+  config.warden do |manager|
+    manager.strategies.add :jwt, Devise::Strategies::JWT
+    manager.default_strategies(scope: :user).unshift :jwt
+    manager.failure_app = JwtFailureApp
+  end
+end
+
+module Devise
```
Comment (omaik) [inline config/initializers/devise.rb]: Move this module to separate file.

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -29,4 +29,44 @@
...
+      private
+
+      def token
+        request.authorization.split.last
+      end
+    end
+  end
+end
+
+class JwtFailureApp < Devise::FailureApp
```
Comment (omaik) [inline config/initializers/devise.rb]: As well as this class

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  post '/signin', to: 'authentication#authenticate'
+
+  devise_for :user, skip: :sessions
```
Comment (omaik) [inline config/routes.rb]: Does this helper add any routes you need?

File: lib/json_web_token.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class JsonWebToken
```
Comment (omaik) [inline lib/json_web_token.rb]: I would change it to module since you don`t create instances of this class

File: config/routes.rb (hunk 3 above)
Comment (vikabelous) [inline config/routes.rb]: @omaik without calling this method it won't add helper methods like `current_user` and `authentiate_user!`, so I added it but skipped all routes

**Comment by omaik** [comment]
(y)

**Pull request merged**